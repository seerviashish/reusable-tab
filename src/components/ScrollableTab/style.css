.tab-container {
  width: 100%;
  min-height: 48px;
  display: flex;
  overflow-y: hidden;
  align-items: center;
}

.tab-container::-webkit-scrollbar {
  display: none;
}

.tab-item {
  margin: 0 2px;
  position: relative;
  text-align: center;
}

.tab-button {
  display: inline-flex;
  width: max-content;
  min-width: 100px;
  justify-content: center;
  padding: 16px 0;
  color: #9e9e9e;
}

.tab-active {
  color: black;
  border-bottom: solid;
  border-bottom-color: #656565;
}

.tab-action-button {
  border: 0;
  margin: 0;
  outline: 0;
  cursor: pointer;
  display: inline-flex;
  align-items: center;
  color: #9e9e9e;
  background: transparent;
}

.tab-bar {
  display: flex;
  background-color: #f3f3f3;
}

.tab-action-button svg {
  fill: #9e9e9e;
}

.light-dark svg {
  fill: rgb(44, 44, 44);
}

.tab-clear-button {
  outline: 0;
  margin: 0 0 0 8px;
  padding: 0;
  border: 0;
  background-color: #f3f3f3;
  cursor: pointer;
  position: relative;
}

.tab-clear-button svg {
  height: 15px;
  width: 15px;
  position: absolute;
  top: -10px;
  fill: #818181;
}
