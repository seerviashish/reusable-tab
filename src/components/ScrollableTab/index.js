import React from "react";
import LeftIcon from "./left";
import RightIcon from "./right";
import AddIcon from "./add";
import ClearIcon from "./clear";
import "./style.css";

const ScrollableTab = (props) => {
  const {
    tabs,
    onSelect,
    onRemove,
    selectedTabIndex,
    onPrev,
    onNext,
    onAdd,
  } = props;

  const scrollItemRef = React.useRef(null);

  const handleOnSelect = (tabId) => (e) => {
    onSelect(tabId);
  };

  const handleRemove = (tabId) => (e) => {
    e.stopPropagation();
    onRemove(tabId);
  };

  const handleScroll = (direction) => (e) => {
    if (direction === "right") {
      if (scrollItemRef && scrollItemRef.current) {
        scrollItemRef.current.scrollLeft += 100;
      }
      onNext();
    } else {
      if (scrollItemRef && scrollItemRef.current) {
        scrollItemRef.current.scrollLeft -= 100;
      }
      onPrev();
    }
  };

  return (
    <div className="tab-bar">
      <button
        className={`tab-action-button ${
          selectedTabIndex === 0 ? "hidden" : ""
        }`}
        onClick={handleScroll("left")}
      >
        <LeftIcon />
      </button>
      <div className="tab-container" ref={scrollItemRef}>
        {tabs.map((tab, index) => {
          return (
            <div
              className="tab-item"
              key={index}
              onClick={handleOnSelect(index)}
            >
              <div
                className={
                  index === selectedTabIndex
                    ? "tab-button tab-active"
                    : "tab-button"
                }
              >
                <span>{tab}</span>
                <button
                  className="tab-clear-button"
                  onClick={handleRemove(index)}
                >
                  <ClearIcon />
                </button>
              </div>
            </div>
          );
        })}
      </div>
      <button
        className={`tab-action-button ${
          selectedTabIndex === tabs.length - 1 ? "hidden" : ""
        }`}
        onClick={handleScroll("right")}
      >
        <RightIcon />
      </button>
      <button className="tab-action-button light-dark" onClick={onAdd}>
        <AddIcon />
      </button>
    </div>
  );
};

export default ScrollableTab;
