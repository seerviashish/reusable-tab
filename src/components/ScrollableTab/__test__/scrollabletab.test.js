import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import AddIcon from "../add";
import ClearIcon from "../clear";
import ScrollableTab from "../index";
import LeftIcon from "../left";
import RightIcon from "../right";
import pretty from "pretty";
import { act } from "@testing-library/react";

let container = null;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("ScrollableTab All Used Icon Component Test", () => {
  it("should match AddIcon snapshot", () => {
    act(() => {
      render(<AddIcon />, container);
    });
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("should match ClearIcon snapshot", () => {
    act(() => {
      render(<ClearIcon />, container);
    });
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("should match LeftIcon snapshot", () => {
    act(() => {
      render(<LeftIcon />, container);
    });
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("should match RightIcon snapshot", () => {
    act(() => {
      render(<RightIcon />, container);
    });
    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});

describe("ScrollableTab Component Test", () => {
  const tabs = ["Tab-1", "Tab-2", "Tab-3"];
  const currentTabIndex = 0;
  const handleDialogOpen = (e) => {};
  const handlePrev = (e) => {};
  const handleNext = (e) => {};
  const handleRemoveTab = (tabIndex) => {};
  const handleOnSelectTab = (tabIndex) => {};

  it("should match ScrollableTab snapshot", () => {
    act(() => {
      render(
        <ScrollableTab
          tabs={tabs}
          selectedTabIndex={currentTabIndex}
          onSelect={handleOnSelectTab}
          onRemove={handleRemoveTab}
          onNext={handleNext}
          onPrev={handlePrev}
          onAdd={handleDialogOpen}
        />,
        container
      );
    });

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});
