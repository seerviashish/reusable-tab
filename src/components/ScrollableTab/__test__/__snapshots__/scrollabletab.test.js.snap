// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollableTab All Used Icon Component Test should match AddIcon snapshot 1`] = `
"<svg xmlns=\\"http://www.w3.org/2000/svg\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" width=\\"24\\">
  <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
  <path d=\\"M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z\\"></path>
</svg>"
`;

exports[`ScrollableTab All Used Icon Component Test should match ClearIcon snapshot 1`] = `
"<svg xmlns=\\"http://www.w3.org/2000/svg\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" width=\\"24\\">
  <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
  <path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path>
</svg>"
`;

exports[`ScrollableTab All Used Icon Component Test should match LeftIcon snapshot 1`] = `
"<svg xmlns=\\"http://www.w3.org/2000/svg\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" width=\\"24\\">
  <path d=\\"M0 0h24v24H0V0z\\" fill=\\"none\\"></path>
  <path d=\\"M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z\\"></path>
</svg>"
`;

exports[`ScrollableTab All Used Icon Component Test should match RightIcon snapshot 1`] = `
"<svg xmlns=\\"http://www.w3.org/2000/svg\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" width=\\"24\\">
  <path d=\\"M0 0h24v24H0V0z\\" fill=\\"none\\"></path>
  <path d=\\"M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z\\"></path>
</svg>"
`;

exports[`ScrollableTab Component Test should match ScrollableTab snapshot 1`] = `
"<div class=\\"tab-bar\\"><button class=\\"tab-action-button hidden\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" width=\\"24\\">
      <path d=\\"M0 0h24v24H0V0z\\" fill=\\"none\\"></path>
      <path d=\\"M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z\\"></path>
    </svg></button>
  <div class=\\"tab-container\\">
    <div class=\\"tab-item\\">
      <div class=\\"tab-button tab-active\\"><span>Tab-1</span><button class=\\"tab-clear-button\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" width=\\"24\\">
            <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
            <path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path>
          </svg></button></div>
    </div>
    <div class=\\"tab-item\\">
      <div class=\\"tab-button\\"><span>Tab-2</span><button class=\\"tab-clear-button\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" width=\\"24\\">
            <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
            <path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path>
          </svg></button></div>
    </div>
    <div class=\\"tab-item\\">
      <div class=\\"tab-button\\"><span>Tab-3</span><button class=\\"tab-clear-button\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" width=\\"24\\">
            <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
            <path d=\\"M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z\\"></path>
          </svg></button></div>
    </div>
  </div><button class=\\"tab-action-button \\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" width=\\"24\\">
      <path d=\\"M0 0h24v24H0V0z\\" fill=\\"none\\"></path>
      <path d=\\"M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z\\"></path>
    </svg></button><button class=\\"tab-action-button light-dark\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" height=\\"24\\" viewBox=\\"0 0 24 24\\" width=\\"24\\">
      <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
      <path d=\\"M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z\\"></path>
    </svg></button>
</div>"
`;
